### PR TITLE
Feature: Add permit to the sdk

### PR DESCRIPTION
**Pull request by brijeshagal**
- Permit 1 and Permit 2 are added.

### CODE REVIEW

File: src/config/index.ts [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
-import { Abi, Chain } from 'viem';
-import { mainnet, arbitrum, bsc, optimism, polygon, zkSync } from 'viem/chains';
 import { AppEnv } from './AppEnv';
 const { REACT_APP_ENV, REACT_APP_BASE_API_URL } = process.env;
 
 let baseUrl = REACT_APP_BASE_API_URL || 'https://api.dzap.io/';
-const stagingUrl = 'https://staging.dzap.io/';
-// const localhostUrl = 'http://localhost:8080/';
+// const stagingUrl = 'https://staging.dzap.io/';
+const stagingUrl = 'http://localhost:8080/';
```
Comment (dzaptech) [inline src/config/index.ts:9]: Always review the code before pushing.

File: src/constants/contract.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import { ContractConfig } from 'src/types/contract';
+import { Services, Versions } from '../enums';
+
+const CHAINS_IDS = {
```
Comment (dzaptech) [inline src/constants/contract.ts:4]: Why do we have this file? Where are we using this?

File: src/constants/contract.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  avalanche: 43114,
+  arbitrum: 42161,
+  optimism: 10,
+  zkSync: 324,
+  base: 8453,
+  scroll: 534352,
+  manta: 169,
+};
+
+export const contractAddress: ContractConfig = {
```
Comment (dzaptech) [inline src/constants/contract.ts:19]: Why are we maintaining contract address in sdk?

File: src/constants/index.ts [hunk 4]
```diff
@@ -11,3 +15,10 @@ export const ERRORS = {
 };
 
 export const HISTORICAL_BLOCK = 10;
+
+export const NATIVE_TOKEN_ADDRESS = '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE';
```
Comment (dzaptech) [inline src/constants/index.ts:19]: We are using 0x as native address at backend, so cross check once.

File: src/contractHandler/index.ts [hunk 5]
```diff
@@ -62,7 +64,7 @@ class ContractHandler {
         wcProjectId: this.wcProjectId,
       });
       return await walletClient.sendTransaction({
-        chain: Chains[chainId],
+        chain: viemchainsById[chainId],
```
Comment (dzaptech) [inline src/contractHandler/index.ts:67]: Chains... C caps

File: src/enums/index.ts [hunk 6]
```diff
@@ -2,3 +2,70 @@ export enum ConnectorType {
...
+
+export enum PermitType {
+  PERMIT,
+  PERMIT2_TRANSFER_FROM,
+  PERMIT2_APPROVE,
+}
+
+export enum TxnStatus {
+  pending = 'pending',
+  inProgress = 'in-progress',
```
Comment (dzaptech) [inline src/enums/index.ts:32]: should ie be inProgress?

File: src/enums/index.ts [hunk 7]
```diff
@@ -2,3 +2,70 @@ export enum ConnectorType {
...
+  success = 'success',
+  rejected = 'rejected',
+  error = 'error',
+  reverted = 'reverted',
+  checkOtherPermit = 'checkOtherPermit',
+}
+
+export enum Versions {
+  V1 = 'v1',
+  V1_2 = 'v1.2',
```
Comment (dzaptech) [inline src/enums/index.ts:42]: There is no 1.2 anymore. Lets check usages and delete unused components.

File: src/utils/index.ts [hunk 8]
```diff
@@ -1,16 +1,16 @@
...
-import { Chains, batchSwapIntegrators, defaultBridgeVersion, defaultSwapVersion } from '../config';
+import { batchSwapIntegrators, defaultBridgeVersion, defaultSwapVersion } from '../config';
 import { HexString } from '../types';
 
 type Window = {
   ethereum: any;
 };
 
-export const wagmiChainsById: Record<number, allWagmiChains.Chain> = Object.values(allWagmiChains).reduce((acc, chainData) => {
+export const viemchainsById: Record<number, allWagmiChains.Chain> = Object.values(allWagmiChains).reduce((acc, chainData) => {
```
Comment (dzaptech) [inline src/utils/index.ts:13]: c caps

File: src/utils/index.ts [hunk 9]
```diff
@@ -66,6 +66,78 @@ export const getWalletClient = async ({
   }
 };
 
+export const readContract = async ({
```
Comment (dzaptech) [inline src/utils/index.ts:69]: this should not be a part of utils... Maybe provideFunctions or viemFunctions

File: src/utils/permit.ts [hunk 10]
```diff
@@ -0,0 +1,167 @@
+import { DEFAULT_PERMIT_DATA, NATIVE_TOKEN_ADDRESS, PERMIT2_APPROVE_DATA } from 'src/constants';
+import { ConnectorType, Erc20Functions, PermitFunctionSelector, PermitSelector, StatusCodes, TxnStatus } from 'src/enums';
+import { HexString } from 'src/types';
+import { erc20Abi } from 'viem';
+
+import { readContract } from './index';
+import { checkPermit1, getPermit1PermitData } from './permit1Methods';
+import { checkPermit2, getPermit2PermitDataForApprove } from './permit2Methods';
+
+export const nativeTokenAndAllowanceChecker = async ({
```
Comment (dzaptech) [inline src/utils/permit.ts:10]: We never check allowance for native token. I know we are returning true but do we need this for native?

File: src/utils/permit1Methods.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+      return { status: TxnStatus.success, code: StatusCodes.Success };
+    }
+    return { status: TxnStatus.checkOtherPermit, code: StatusCodes.FunctionNotFound };
+  } catch (e) {
+    console.log({ e });
+    return { status: TxnStatus.checkOtherPermit, code: StatusCodes.FunctionNotFound };
+  }
+};
+
+export const getPermit1PermitData = async ({
```
Comment (dzaptech) [inline src/utils/permit1Methods.ts:44]: getPermit1Data ?? instead of getPermit1PermitData

File: src/utils/permit2Methods.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  } catch (e) {
+    console.log({ e });
+    if (e?.cause?.code === StatusCodes.UserRejectedRequest || e?.code === StatusCodes.UserRejectedRequest) {
+      return { status: TxnStatus.rejected, errorCode: StatusCodes.UserRejectedRequest };
+    }
+    return { status: TxnStatus.error, code: e.code };
+  }
+};
+
+export async function getPermit2PermitDataForApprove({
```
Comment (dzaptech) [inline src/utils/permit2Methods.ts:69]: getPermit2DataForApprove ??

File: src/constants/index.ts (hunk 4 above)
Comment (brijeshagal) [reply src/constants/index.ts:19]: I had asked @dante055 about this. Will cross check and update again.

File: src/utils/permit.ts (hunk 10 above)
Comment (brijeshagal) [reply src/utils/permit.ts:10]: This is the case of nativeToken or Allowance is already given case where permit would have the same default value.

File: src/utils/permit1Methods.ts (hunk 11 above)
Comment (brijeshagal) [reply src/utils/permit1Methods.ts:44]: Because we return permit data, so to keep it explicit.

**Pull request closed without merging**